src/backend/gl/src/lib.rs

```diff
@@ -481,37 +495,68 @@ impl hal::PhysicalDevice<Backend> for PhysicalDevice {
     fn memory_properties(&self) -> hal::MemoryProperties {
         use crate::hal::memory::Properties;
 
-        // COHERENT flags require that the backend does flushing and invalidation
-        // by itself. If we move towards persistent mapping we need to re-evaluate it.
         let caps = &self.0.private_caps;
-        let memory_types = if caps.map || caps.emulate_map {
-            vec![
+        assert!(caps.map || caps.emulate_map);
+
+        // For security reasons, WebGL does not allow "element array buffers" to be used as any
+        // other kind of buffer.  We need to provide unique types of memory specifically for buffers
+        // with INDEX usage.
+        let mut memory_types = vec![
+            // Faked DEVICE_LOCAL memory for images, no gl buffer is actually allocated for them.
+            hal::MemoryType {
+                properties: Properties::DEVICE_LOCAL,
+                heap_index: 0,
+            },
+            // DEVICE_LOCAL memory for INDEX buffers
+            hal::MemoryType {
+                properties: Properties::DEVICE_LOCAL,
+                heap_index: 0,
+            },
+            // DEVICE_LOCAL memory for non-INDEX buffers
+            hal::MemoryType {
+                properties: Properties::DEVICE_LOCAL,
+                heap_index: 0,
+            },
+            // Incoherent CPU_VISIBLE memory for INDEX buffers
+            hal::MemoryType {
+                properties: Properties::CPU_VISIBLE
+                    | Properties::CPU_CACHED,
+                heap_index: 1,
+            },
+            // Incoherent CPU_VISIBLE memory for non-INDEX buffers
+            hal::MemoryType {
+                properties: Properties::CPU_VISIBLE
+                    | Properties::CPU_CACHED,
+                heap_index: 1,
+            },
+        ];
+
+        // If we are not emulating buffer mapping, we can provide coherent buffer mapping by
+        // omitting the GL_MAP_FLUSH_EXPLICIT_BIT flag.
+        if !self.0.private_caps.emulate_map {
+            memory_types.push(
+                // Coherent CPU_VISIBLE memory for INDEX buffers
```

Comment:
let's move these to the beginning of `memory_types`. In Vulkan there is an interesting requirement that a memory type with more flags should be exposed first (i.e one with flags A | B should come before A).

Reply:
That makes the index masks dynamic based on capabilities, darn.